### PR TITLE
feat: Make the download button download all files

### PR DESCRIPTION
**Pull request by LouisCAD**
Before this PR, the button was no-op.

### CODE REVIEW

File: Core2/src/main/java/com/infomaniak/core2/DownloadManagerUtils.kt [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+    private val regexInvalidSystemChar = Regex("[\\\\/:*?\"<>|\\x7F]|[\\x00-\\x1f]")
+
+    private val scope = CoroutineScope(Dispatchers.Default)
+
+    fun scheduleDownload(
+        url: String,
+        name: String,
+        userAgent: String,
+        extraHeaders: Iterable<Pair<String, String>> = emptySet()
```
Comment (FabianDevel) [inline Core2/src/main/java/com/infomaniak/core2/DownloadManagerUtils.kt:43]:
```suggestion
        extraHeaders: Iterable<Pair<String, String>> = emptySet(),
```

File: Core2/src/main/java/com/infomaniak/core2/DownloadManagerUtils.kt [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+            setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
+
+            val downloadManager = appCtx.getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager
+            val downloadReference = downloadManager.enqueue(this)
+            handleDownloadManagerErrors(downloadManager, downloadReference)
+        }
+    }
+
+    private fun DownloadManager.Request.addHeaders(userAgent: String, extraHeaders: Iterable<Pair<String, String>>) {
```
Comment (FabianDevel) [inline Core2/src/main/java/com/infomaniak/core2/DownloadManagerUtils.kt:65]: Import DownloadManager here to only keep one level of `parenting`

File: Core2/src/main/java/com/infomaniak/core2/DownloadManagerUtils.kt [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        addRequestHeader("Accept-Encoding", "gzip")
+        addRequestHeader("App-Version", "Android $appVersionName")
+        addRequestHeader("User-Agent", userAgent)
+        extraHeaders.forEach { (key, value) ->
+            addRequestHeader(key, value)
+        }
```
Comment (FabianDevel) [inline Core2/src/main/java/com/infomaniak/core2/DownloadManagerUtils.kt:71]:
```suggestion
        extraHeaders.forEach { (key, value) -> addRequestHeader(key, value)  }
```

File: Core2/src/main/java/com/infomaniak/core2/DownloadManagerUtils.kt [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    private fun handleDownloadManagerErrors(downloadManager: DownloadManager, downloadReference: Long) {
+        scope.launch {
+            delay(1_000L)
```
Comment (FabianDevel) [inline Core2/src/main/java/com/infomaniak/core2/DownloadManagerUtils.kt:76]: Was this delay for testing purpose or is it necessary ? (and why if it's the case )

File: Core2/src/main/java/com/infomaniak/core2/DownloadManagerUtils.kt [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+                }
+            }
+        }
+    }
+
+    @OptIn(UnreliableToastApi::class)
+    private suspend fun showErrorIfAny(cursor: Cursor) {
+        val status: Int
+        val reason: Int
+        withContext(Dispatchers.IO) {
```
Comment (FabianDevel) [inline Core2/src/main/java/com/infomaniak/core2/DownloadManagerUtils.kt:92]:
```suggestion
        val reason: Int
        
        withContext(Dispatchers.IO) {
```

File: Core2/src/main/java/com/infomaniak/core2/extensions/PackageManager.kt [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+val appName: String by lazy(NONE) {
+    packageManager.getApplicationLabel(applicationInfo).toString()
+}
+
+val appVersionName: String by lazy(NONE) {
+    packageManager.getPackageInfo(appCtx.packageName, 0).versionName.toString()
+}
+
+private val packageManager = appCtx.packageManager
+private val applicationInfo by lazy(NONE) { appCtx.applicationInfo }
```
Comment (FabianDevel) [inline Core2/src/main/java/com/infomaniak/core2/extensions/PackageManager.kt:32]: We always keep one blankline at the end of files
You can add it to your IDE parameters so it makes it automatically

File: app/src/main/java/com/infomaniak/swisstransfer/ui/screen/main/transferdetails/TransferDetailsViewModel.kt [hunk 7]
```diff
@@ -69,6 +72,34 @@ class TransferDetailsViewModel @Inject constructor(
...
 
+    fun startDownloadingAllFiles(transfer: TransferUi) {
+        viewModelScope.launch {
+            val url = sharedApiUrlCreator.downloadFilesUrl(transfer.uuid) ?: return@launch
+            val userAgent = buildUserAgent(
+                appId = BuildConfig.APPLICATION_ID,
+                appVersionCode = BuildConfig.VERSION_CODE,
+                appVersionName = BuildConfig.VERSION_NAME,
+            )
+            when (transfer.files.size) {
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/swisstransfer/ui/screen/main/transferdetails/TransferDetailsViewModel.kt:83]:
```suggestion
            )
            
            when (transfer.files.size) {
```

File: app/src/main/java/com/infomaniak/swisstransfer/ui/screen/main/transferdetails/TransferDetailsViewModel.kt [hunk 8]
```diff
@@ -69,6 +72,34 @@ class TransferDetailsViewModel @Inject constructor(
...
+                        url = sharedApiUrlCreator.downloadFileUrl(transfer.uuid, singleFile.uid) ?: return@launch,
+                        name = "SwissTransfer/${singleFile.fileName}",
+                        userAgent = userAgent
+                    )
+                }
+                else -> {
+                    DownloadManagerUtils.scheduleDownload(
+                        url = url,
+                        name = "SwissTransfer/${transfer.uuid}.zip", //TODO: Prepend date (and ensure the iOS app does it too)
+                        userAgent = userAgent
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/swisstransfer/ui/screen/main/transferdetails/TransferDetailsViewModel.kt:96]:
```suggestion
                        userAgent = userAgent,
```

File: app/src/main/java/com/infomaniak/swisstransfer/ui/screen/main/transferdetails/TransferDetailsViewModel.kt [hunk 9]
```diff
@@ -69,6 +72,34 @@ class TransferDetailsViewModel @Inject constructor(
...
+                appVersionCode = BuildConfig.VERSION_CODE,
+                appVersionName = BuildConfig.VERSION_NAME,
+            )
+            when (transfer.files.size) {
+                1 -> {
+                    val singleFile = transfer.files.single()
+                    DownloadManagerUtils.scheduleDownload(
+                        url = sharedApiUrlCreator.downloadFileUrl(transfer.uuid, singleFile.uid) ?: return@launch,
+                        name = "SwissTransfer/${singleFile.fileName}",
+                        userAgent = userAgent
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/swisstransfer/ui/screen/main/transferdetails/TransferDetailsViewModel.kt:89]:
```suggestion
                        userAgent = userAgent,
```

File: app/src/main/java/com/infomaniak/swisstransfer/ui/screen/main/transferdetails/TransferDetailsViewModel.kt [hunk 10]
```diff
@@ -69,6 +72,34 @@ class TransferDetailsViewModel @Inject constructor(
...
+            val userAgent = buildUserAgent(
+                appId = BuildConfig.APPLICATION_ID,
+                appVersionCode = BuildConfig.VERSION_CODE,
+                appVersionName = BuildConfig.VERSION_NAME,
+            )
+            when (transfer.files.size) {
+                1 -> {
+                    val singleFile = transfer.files.single()
+                    DownloadManagerUtils.scheduleDownload(
+                        url = sharedApiUrlCreator.downloadFileUrl(transfer.uuid, singleFile.uid) ?: return@launch,
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/swisstransfer/ui/screen/main/transferdetails/TransferDetailsViewModel.kt:87]: Maybe we shouldn't fail silently here, and show a snackbar to the user that it didn't work

**Review by FabianDevel — CHANGES_REQUESTED** [review]
I only quickly checked the code formating
As I won't be here for 2 weeks to manage the aftermath of my comments, I prefer to let the logic be reviewed by someone who will be here during christmas time
This way I won't be blocking the PR

File: Core2/src/main/java/com/infomaniak/core2/DownloadManagerUtils.kt (hunk 4 above)
Comment (sirambd) [reply Core2/src/main/java/com/infomaniak/core2/DownloadManagerUtils.kt]: This is to make sure you get the errors from the query

File: app/src/main/java/com/infomaniak/swisstransfer/ui/screen/main/transferdetails/TransferDetailsViewModel.kt (hunk 10 above)
Comment (sirambd) [reply app/src/main/java/com/infomaniak/swisstransfer/ui/screen/main/transferdetails/TransferDetailsViewModel.kt:87]: it'll be change In an upcoming PR that depends on a new KMP release

**Review by sirambd — APPROVED** [review]

**Review by FabianDevel — APPROVED** [review]

**Pull request merged**